gems/aws-sdk-core/lib/seahorse/client/plugins/request_callback.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+require 'pathname'
+require 'forwardable'
+
+module Seahorse
+  module Client
+    module Plugins
+
+      # @api private
+      class ReadCallbackIO
+        extend Forwardable
+        def_delegators :@io, :size
+
+        def initialize(io, on_read = nil)
+          @io = io
+          @on_read = on_read if on_read.is_a? Proc
+          @bytes_read = 0
+        end
+
+        attr_reader :io
+
+        def read(*args)
+          ret = @io.read(*args)
+          @bytes_read += ret.bytesize if ret && ret.respond_to?(:bytesize)
+          total_size = @io.respond_to?(:size) ? @io.size : nil
+          @on_read.call(ret, @bytes_read, total_size) if @on_read
+          ret
+        end
+      end
+
+      # @api private
+      class RequestCallback < Plugin
+
+        option(:on_chunk_sent,
+               default: nil,
+               doc_type: 'Proc',
+               docstring: <<-DOCS)
+When a Proc object is provided, it will be used as callback when each chunk 
+of the request body is sent. It will be called with three arguments: the chunk,
+the number, of bytes read from the body, and the total number of 
+bytes in the body.
+          DOCS
+
+        # @api private
+        class OptionHandler < Client::Handler
+          def call(context)
+            if context.params.is_a?(Hash)
```

Comment:
Never in the case of a service client.  However, since this plugin is in Seahorse rather than AWS, its possible (see the ResponseTarget plugin).

Reply:
Just updating this for posterity - this is actually not because its a Seahorse plugin.  This check is useful because the `OptionHandler` is in the `:initialize` step, which happens before `:validate`.  If someone calls `client.operation('not_a_hash')` then `params` will be a string rather than a hash at this point.  This is not a valid call, and the `ParamValidator` will raise a user friendly `ArgumentError`.